File: base/src/main/vsdk/toolkit/media/RGBAImageCompressed.java
package vsdk.toolkit.media;
import java.io.Serial;
import java.nio.ByteBuffer;
import java.util.Arrays;

import vsdk.toolkit.common.VSDK;

/**
Compressed RGBA image storage. Pixel-level access is intentionally unsupported
because the contents remain in GPU-ready compressed blocks.
*/
public class RGBAImageCompressed extends Image
{
    @SuppressWarnings("FieldNameHidesFieldInSuperclass")
    @Serial private static final long serialVersionUID = 20260430L;

    public static final int COMPRESSION_UNKNOWN = 0;
    public static final int COMPRESSION_DXT1 = 1;
    public static final int COMPRESSION_DXT3 = 3;
    public static final int COMPRESSION_DXT5 = 5;

    private byte data[];
    transient private ByteBuffer directData;
    private int xSize;
    private int ySize;
    private int compressionFormat;
    private int compressedDataSize;

    public RGBAImageCompressed()
    {
        data = null;
        directData = null;
        xSize = 0;
        ySize = 0;
        compressionFormat = COMPRESSION_UNKNOWN;
        compressedDataSize = 0;
    }

    @Override
    public void finalize()
    {
        if ( data != null ) {
            data = null;
            directData = null;
            xSize = 0;
            ySize = 0;
            compressionFormat = COMPRESSION_UNKNOWN;
            compressedDataSize = 0;
        }
        try {
            super.finalize();
        }
        catch (Throwable ex) {
        }
    }

    public void dettach()
    {
        data = null;
        directData = null;
    }

    @Override
    public int getSizeInBytes()
    {
        int dataSize = data == null ? 0 : data.length;
        return dataSize + 4*INT_SIZE_IN_BYTES + POINTER_SIZE_IN_BYTES;
    }

    @Override
    public boolean init(int width, int height)
    {
        xSize = width;
        ySize = height;
        compressionFormat = COMPRESSION_UNKNOWN;
        compressedDataSize = 0;
        data = new byte[0];
        directData = ByteBuffer.allocateDirect(0);
        return true;
    }

    @Override
    public boolean initNoFill(int width, int height)
    {
        return init(width, height);
    }

    public boolean initCompressed(
        int width,
        int height,
        int compressionFormat,
        byte compressedData[])
    {
        if ( width <= 0 || height <= 0 || compressedData == null ) {
            data = null;
            directData = null;
            xSize = 0;
            ySize = 0;
            this.compressionFormat = COMPRESSION_UNKNOWN;
            compressedDataSize = 0;
            return false;
        }

        xSize = width;
        ySize = height;
        this.compressionFormat = compressionFormat;
        compressedDataSize = calculateTopLevelDataSize(width, height, compressionFormat);
        if ( compressedData.length < compressedDataSize ) {
            data = null;
            directData = null;
            xSize = 0;
            ySize = 0;
            this.compressionFormat = COMPRESSION_UNKNOWN;
            compressedDataSize = 0;
            return false;
        }
        data = Arrays.copyOf(compressedData, compressedData.length);
        directData = ByteBuffer.allocateDirect(data.length);
        directData.put(data);
        directData.rewind();
        return true;
    }

    public void setRawImage(
        int width,
        int height,
        int compressionFormat,
        byte compressedData[])
    {
        initCompressed(width, height, compressionFormat, compressedData);
    }

    public int getCompressionFormat()
    {
        return compressionFormat;
    }

    public int getCompressedDataSize()
    {
        return compressedDataSize;
    }

    public byte[] getRawImage()
    {
        return data;
    }

    public ByteBuffer getRawImageDirectBuffer()
    {
        directData.rewind();
        return directData;
    }

    @Override
    public int getXSize()
    {
        return xSize;
    }

    @Override
    public int getYSize()
    {
        return ySize;
    }

    public void putPixel(int x, int y, byte r, byte g, byte b)
    {
        reportUnsupportedPixelAccess("putPixel");
    }

    public void putPixel(int x, int y, byte r, byte g, byte b, byte a)
    {
        reportUnsupportedPixelAccess("putPixel");
    }

    public void putPixel(int x, int y, RGBAPixel p)
    {
        reportUnsupportedPixelAccess("putPixel");
    }

    @Override
    public void putPixelRgb(int x, int y, RGBPixel p)
    {
        reportUnsupportedPixelAccess("putPixelRgb");
    }

    public RGBAPixel getPixel(int x, int y)
    {
        reportUnsupportedPixelAccess("getPixel");
        return new RGBAPixel();
    }

    @Override
    public RGBPixel getPixelRgb(int x, int y)
    {
        reportUnsupportedPixelAccess("getPixelRgb");
        return new RGBPixel();
    }

    public void getPixelRgba(int x, int y, RGBAPixel p)
    {
        reportUnsupportedPixelAccess("getPixelRgba");
    }

    @Override
    public void getPixelRgb(int x, int y, RGBPixel p)
    {
        reportUnsupportedPixelAccess("getPixelRgb");
    }

    @Override
    public RGBAImageCompressed clone() throws CloneNotSupportedException
    {
        super.clone();
        RGBAImageCompressed copy = new RGBAImageCompressed();
        copy.initCompressed(xSize, ySize, compressionFormat, data);
        return copy;
    }

    public static int calculateTopLevelDataSize(
        int width,
        int height,
        int compressionFormat)
    {
        int blockSize;

        if ( compressionFormat == COMPRESSION_DXT1 ) {
            blockSize = 8;
        }
        else if ( compressionFormat == COMPRESSION_DXT3 ||
                  compressionFormat == COMPRESSION_DXT5 ) {
            blockSize = 16;
        }
        else {
            return 0;
        }

        int blockWidth = Math.max(1, (width + 3) / 4);
        int blockHeight = Math.max(1, (height + 3) / 4);
        return blockWidth * blockHeight * blockSize;
    }

    private void reportUnsupportedPixelAccess(String method)
    {
        VSDK.reportMessage(
            this,
            VSDK.FATAL_ERROR,
            method,
            "Pixel access is not implemented for RGBAImageCompressed.");
    }
}

File: cpp/base/src/main/vsdk/toolkit/media/RGBAImageCompressed.cpp
#include "vsdk/toolkit/media/RGBAImageCompressed.h"
#include "vsdk/toolkit/media/RGBPixel.h"
#include "vsdk/toolkit/common/VSDK.h"
#include "vsdk/toolkit/common/logging/Logger.h"
#include <cstring>
#include <algorithm>

RGBAImageCompressed::RGBAImageCompressed() :
    data(nullptr), xSize(0), ySize(0),
    compressionFormat(COMPRESSION_UNKNOWN), compressedDataSize(0) {
}

RGBAImageCompressed::~RGBAImageCompressed() {
    if (data != nullptr) {
        delete[] data;
        data = nullptr;
    }
}

void RGBAImageCompressed::detach() {
    if (data != nullptr) {
        delete[] data;
        data = nullptr;
    }
}

int RGBAImageCompressed::getSizeInBytes() const {
    int dataSize = (data == nullptr) ? 0 : compressedDataSize;
    return dataSize + 4 * INT_SIZE_IN_BYTES + POINTER_SIZE_IN_BYTES;
}

bool RGBAImageCompressed::init(int width, int height) {
    xSize = width;
    ySize = height;
    compressionFormat = COMPRESSION_UNKNOWN;
    compressedDataSize = 0;
    if (data != nullptr) {
        delete[] data;
    }
    data = new char[0];
    return true;
}

bool RGBAImageCompressed::initNoFill(int width, int height) {
    return init(width, height);
}

int RGBAImageCompressed::calculateTopLevelDataSize(
    int width,
    int height,
    int compressionFormat) {
    int blockSize;

    if (compressionFormat == COMPRESSION_DXT1) {
        blockSize = 8;
    } else if (compressionFormat == COMPRESSION_DXT3 ||
               compressionFormat == COMPRESSION_DXT5) {
        blockSize = 16;
    } else {
        return 0;
    }

    int blockWidth = std::max(1, (width + 3) / 4);
    int blockHeight = std::max(1, (height + 3) / 4);
    return blockWidth * blockHeight * blockSize;
}

bool RGBAImageCompressed::initCompressed(
    int width,
    int height,
    int compressionFormat_,
    char* compressedData,
    int dataSize) {
    if (width <= 0 || height <= 0 || compressedData == nullptr) {
        data = nullptr;
        xSize = 0;
        ySize = 0;
        compressionFormat = COMPRESSION_UNKNOWN;
        compressedDataSize = 0;
        return false;
    }

    xSize = width;
    ySize = height;
    compressionFormat = compressionFormat_;
    compressedDataSize = calculateTopLevelDataSize(width, height, compressionFormat_);
    if (dataSize < compressedDataSize) {
        data = nullptr;
        xSize = 0;
        ySize = 0;
        compressionFormat = COMPRESSION_UNKNOWN;
        compressedDataSize = 0;
        return false;
    }

    if (data != nullptr) {
        delete[] data;
    }
    data = new char[dataSize];
    std::memcpy(data, compressedData, dataSize);
    return true;
}

void RGBAImageCompressed::setRawImage(
    int width,
    int height,
    int compressionFormat_,
    char* compressedData,
    int dataSize) {
    initCompressed(width, height, compressionFormat_, compressedData, dataSize);
}

int RGBAImageCompressed::getCompressionFormat() const {
    return compressionFormat;
}

int RGBAImageCompressed::getCompressedDataSize() const {
    return compressedDataSize;
}

char* RGBAImageCompressed::getRawImage() const {
    if (data == nullptr) {
        return nullptr;
    }
    char* copy = new char[compressedDataSize];
    std::memcpy(copy, data, compressedDataSize);
    return copy;
}

const char* RGBAImageCompressed::getRawImageDirectBuffer() const {
    return data;
}

void RGBAImageCompressed::reportUnsupportedPixelAccess(const char* method) const {
    fprintf(stderr, "FATAL_ERROR in %s: Pixel access is not implemented for RGBAImageCompressed.\n", method);
}

void RGBAImageCompressed::putPixelRgb(int x, int y, RGBPixel* p) {
    reportUnsupportedPixelAccess("putPixelRgb");
}

RGBPixel* RGBAImageCompressed::getPixelRgb(int x, int y) const {
    reportUnsupportedPixelAccess("getPixelRgb");
    return new RGBPixel();
}

void RGBAImageCompressed::getPixelRgb(int x, int y, RGBPixel* p) const {
    reportUnsupportedPixelAccess("getPixelRgb");
}

int RGBAImageCompressed::getXSize() const {
    return xSize;
}

int RGBAImageCompressed::getYSize() const {
    return ySize;
}
